### PR TITLE
Meedan 8403 suggested similarity

### PR DESCRIPTION
**Pull request by DGaffney**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -49,10 +49,12 @@ rvm:
...
   include:
+#    - stage: test
+#      script: "export PATTERN='models mailers integration workers lib' && cd test && rm -rf $PATTERN && echo 'Running tests:' && ls && cd - && bundle exec rake parallel:test[5] && cd test && ./format-coverage.sh && cd -"
+#    - stage: test
+#      script: "export PATTERN='controllers' && cd test && rm -rf $PATTERN && echo 'Running tests:' && ls && cd - && bundle exec rake parallel:test[5] && cd test && ./format-coverage.sh && cd -"
     - stage: test
-      script: "export PATTERN='models mailers integration workers lib' && cd test && rm -rf $PATTERN && echo 'Running tests:' && ls && cd - && bundle exec rake parallel:test[5] && cd test && ./format-coverage.sh && cd -"
-    - stage: test
-      script: "export PATTERN='controllers' && cd test && rm -rf $PATTERN && echo 'Running tests:' && ls && cd - && bundle exec rake parallel:test[5] && cd test && ./format-coverage.sh && cd -"
+      script: "bundle exec ruby test/models/bot/alegre_test.rb && bundle exec ruby test/models/project_media_test.rb -n /falsity/"
```
Comment (caiosba) [inline .travis.yml:57]: Please revert this change

Comment (DGaffney) [reply .travis.yml:57]: I think we discussed that tests are no longer necessary - joke, will revert!

File: app/graph/types/project_media_type.rb [hunk 2]
```diff
@@ -284,5 +284,13 @@
     }
   end
 
+  connection :suggested_similar_items, -> { ProjectMediaType.connection_type } do
+    resolve -> (project_media, args, _ctx) {
```
Comment (caiosba) [inline app/graph/types/project_media_type.rb:288]: We're not using `args` in this method, so please rename it to `_args` (convention to mark parameters that are not used)

Comment (DGaffney) [reply app/graph/types/project_media_type.rb:288]: Is that what CodeClimate is going off about?

File: config/config.yml.example [hunk 3]
```diff
@@ -22,6 +22,11 @@ development: &default
   # '(https?://.*\.?(meedan.com|meedan.org))'
   allowed_origins: '.*'
 
+  # Alegre Similarity Thresholds
+  #
+  image_similarity_threshold: 0.7
+  text_similarity_threshold: 0.7
+
```
Comment (caiosba) [inline config/config.yml.example:29]: Please mark whether those configuration keys are mandatory or optional and what they are used for, like we do with other configuration keys.

File: config/config.yml.example [hunk 4]
```diff
@@ -227,6 +232,11 @@ test:
     bucket: 'check-api-test'
     bucket_region: 'us-east-1'
 
+  # Alegre Similarity Thresholds
+  #
+  image_similarity_threshold: 0.7
+  text_similarity_threshold: 0.7
+
```
Comment (caiosba) [inline config/config.yml.example:239]: We don't need to repeat the configuration keys in the `test` section if they are the same as the `development` section.

File: test/models/bot/alegre_test.rb [hunk 5]
```diff
@@ -23,6 +23,12 @@ def setup
           'confidence': 1.0
         }
       }.to_json)
+      Bot::Alegre.stubs(:request_api).returns({
+        'result' => {
+          'language' => 'en',
+          'confidence' => 1.0
+        }
+      })
```
Comment (caiosba) [inline test/models/bot/alegre_test.rb:31]: Please make sure to `unstub` at the end of the test (or at once in `def teardown`) because it doesn't happen automatically in this version of the library. Something like: `Bot::Alegre.unstub(:request_api)`.

File: test/models/bot/alegre_test.rb [hunk 6]
```diff
@@ -35,6 +41,7 @@ def setup
       WebMock.stub_request(:get, 'http://alegre/text/langid/').to_return(body: {
         'foo': 'bar'
       }.to_json)
+      Bot::Alegre.stubs(:request_api).raises(RuntimeError)
```
Comment (caiosba) [inline test/models/bot/alegre_test.rb:44]: Please make sure to `unstub` at the end of the test (or at once in `def teardown`) because it doesn't happen automatically in this version of the library. Something like: `Bot::Alegre.unstub(:request_api)`.

File: test/models/bot/alegre_test.rb [hunk 7]
```diff
@@ -61,6 +68,7 @@ def setup
       }.to_json)
       WebMock.stub_request(:post, 'http://alegre/image/similarity/').to_return(body: 'success')
       pm1 = create_project_media team: @pm.team, media: create_uploaded_image
+      Bot::Alegre.stubs(:media_file_url).with(pm1).returns("some/path")
```
Comment (caiosba) [inline test/models/bot/alegre_test.rb:71]: Please make sure to `unstub` at the end of the test (or at once in `def teardown`) because it doesn't happen automatically in this version of the library. Something like: `Bot::Alegre.unstub(:request_api)`.

File: test/models/bot/alegre_test.rb [hunk 8]
```diff
@@ -79,12 +87,15 @@ def setup
         ]
       }.to_json)
       pm2 = create_project_media team: @pm.team, media: create_uploaded_image
-      assert_equal [pm1.id], Bot::Alegre.get_items_with_similar_image(pm2, 0.9)
+      response = {pm1.id => 0}
+      Bot::Alegre.stubs(:media_file_url).with(pm2).returns("some/path")
```
Comment (caiosba) [inline test/models/bot/alegre_test.rb:91]: Please make sure to `unstub` at the end of the test (or at once in `def teardown`) because it doesn't happen automatically in this version of the library. Something like: `Bot::Alegre.unstub(:request_api)`.

File: test/models/bot/alegre_test.rb [hunk 9]
```diff
@@ -79,12 +87,15 @@ def setup
...
+      assert_equal response, Bot::Alegre.get_items_with_similar_image(pm2, 0.9)
       assert_nil pm2.get_annotations('flag').last
       WebMock.stub_request(:get, 'http://alegre/image/classification/').to_return(body: {
         "result": valid_flags_data
       }.to_json)
       pm3 = create_project_media team: @pm.team, media: create_uploaded_image
+      Bot::Alegre.stubs(:media_file_url).with(pm3).returns("some/path")
```
Comment (caiosba) [inline test/models/bot/alegre_test.rb:98]: Please make sure to `unstub` at the end of the test (or at once in `def teardown`) because it doesn't happen automatically in this version of the library. Something like: `Bot::Alegre.unstub(:request_api)`.

File: test/models/bot/alegre_test.rb [hunk 10]
```diff
@@ -99,6 +110,12 @@ def setup
           'confidence': 1.0
         }
       }.to_json)
+      Bot::Alegre.stubs(:request_api).returns({
+        'result' => {
+          'language' => 'en',
+          'confidence' => 1.0
+        }
+      })
```
Comment (caiosba) [inline test/models/bot/alegre_test.rb:118]: Please make sure to `unstub` at the end of the test (or at once in `def teardown`) because it doesn't happen automatically in this version of the library. Something like: `Bot::Alegre.unstub(:request_api)`.

File: test/models/bot/alegre_test.rb [hunk 11]
```diff
@@ -114,16 +131,74 @@ def setup
...
-    pm1 = create_project_media project: p
-    pm2 = create_project_media project: p
-    pm3 = create_project_media project: p
-    create_relationship source_id: pm3.id, target_id: pm2.id
-    Bot::Alegre.add_relationships(pm1, [pm2.id])
+    pm1 = create_project_media project: p, is_image: true
+    pm2 = create_project_media project: p, is_image: true
+    pm3 = create_project_media project: p, is_image: true
+    create_relationship source_id: pm2.id, target_id: pm1.id
+    Bot::Alegre.stubs(:request_api).returns({
```
Comment (caiosba) [inline test/models/bot/alegre_test.rb:158]: Please make sure to `unstub` at the end of the test (or at once in `def teardown`) because it doesn't happen automatically in this version of the library. Something like: `Bot::Alegre.unstub(:request_api)`.

File: test/models/bot/alegre_test.rb [hunk 12]
```diff
@@ -114,16 +131,74 @@ def setup
...
+          "url" => "https:\/\/www.gstatic.com\/webp\/gallery3\/1.png",
+          "context" => [{
+            "team_id" => pm2.team.id.to_s,
+            "project_media_id" => pm2.id.to_s
+          }],
+          "score" => 1.0
+        }
+      ]
+    })
+    Bot::Alegre.stubs(:media_file_url).with(pm3).returns("some/path")
```
Comment (caiosba) [inline test/models/bot/alegre_test.rb:173]: Please make sure to `unstub` at the end of the test (or at once in `def teardown`) because it doesn't happen automatically in this version of the library. Something like: `Bot::Alegre.unstub(:request_api)`.

File: app/graph/types/project_media_type.rb (hunk 2 above)
Comment (caiosba) [reply app/graph/types/project_media_type.rb:288]: Yes! :D

**Comment by caiosba** [comment]
@DGaffney please make sure to run all tests now

File: test/controllers/graphql_controller_3_test.rb [hunk 13]
```diff
@@ -1589,6 +1589,27 @@ def setup
     assert_match /Sorry/, @response.body
   end
 
+
+test "should return suggested similar items" do
```
Comment (caiosba) [inline test/controllers/graphql_controller_3_test.rb:1593]: Can you please fix the indentation here? :)

File: test/models/ability_test.rb [hunk 14]
```diff
@@ -2073,7 +2073,7 @@ def teardown
     end
   end
 
-  test "should not duplicate query conditions" do
+  test "zzz should not duplicate query conditions" do
```
Comment (caiosba) [inline test/models/ability_test.rb:2076]: Please remove the :zzz: :D

File: app/models/bot/alegre.rb [hunk 15]
```diff
@@ -123,56 +148,78 @@ def self.request_api(method, path, params = {})
...
   end
 
-  def self.get_items_with_similar_title(pm, threshold)
-    self.get_items_with_similar_text(pm, 'title', threshold, pm.title)
+  def self.get_items_with_similar_title(pm, threshold, text_length_threshold=CONFIG["similarity_text_length_threshold"])
+    pm.title.to_s.split(" ").length > text_length_threshold ? self.get_items_with_similar_text(pm, 'title', threshold, pm.title) : {}
+  end
+
+  def self.get_items_with_similar_description(pm, threshold, text_length_threshold=CONFIG["similarity_text_length_threshold"])
+    pm.description.to_s.split(" ").length > text_length_threshold ? self.get_items_with_similar_text(pm, 'description', threshold, pm.description) : {}
```
Comment (caiosba) [inline app/models/bot/alegre.rb:156]: I think it's more reliable to split by `\s` than a single space.

File: app/models/bot/alegre.rb [hunk 16]
```diff
@@ -123,56 +148,78 @@ def self.request_api(method, path, params = {})
     end
   end
 
-  def self.get_items_with_similar_title(pm, threshold)
-    self.get_items_with_similar_text(pm, 'title', threshold, pm.title)
+  def self.get_items_with_similar_title(pm, threshold, text_length_threshold=CONFIG["similarity_text_length_threshold"])
+    pm.title.to_s.split(" ").length > text_length_threshold ? self.get_items_with_similar_text(pm, 'title', threshold, pm.title) : {}
```
Comment (caiosba) [inline app/models/bot/alegre.rb:152]: I think it's more reliable to split by \s than a single space.

**Review by caiosba — CHANGES_REQUESTED** [review]

File: app/models/bot/alegre.rb [hunk 17]
```diff
@@ -20,6 +19,7 @@ def self.run(body)
         self.send_to_image_similarity_index(pm)
         self.send_to_title_similarity_index(pm)
```
Comment (caiosba) [inline app/models/bot/alegre.rb:20]: Need to send the description to the index too. I see we're still only sending the title, although you do search for title and description.

File: app/models/project_media.rb [hunk 18]
```diff
@@ -32,6 +32,34 @@ class ProjectMedia < ActiveRecord::Base
...
+    self.media_type == "Link"
+  end
+
+  def is_uploaded_image?
+    self.media_type == "UploadedImage"
+  end
+
+  def is_blank?
+    self.media_type == "Blank"
+  end
```
Comment (caiosba) [inline app/models/project_media.rb:53]: These methods need to be fixed: `media_type` is a setter for the `ProjectMedia` model which is persisted only in memory while the item is being created, so it will be `nil` for database-persisted data. Need to replace `media_type` by `media.type` in all these methods.

**Comment by caiosba** [comment]
I merged this branch with latest `develop` and this test is failing:
```
Bot::AlegreTest#test_should_set_user_id_on_relationships:
NoMethodError: undefined method `keys' for [58]:Array
    app/models/bot/alegre.rb:213:in `add_relationships'
    test/models/bot/alegre_test.rb:426:in `block in <class:AlegreTest>'
```

File: test/controllers/graphql_controller_3_test.rb (hunk 13 above)
Comment (DGaffney) [reply test/controllers/graphql_controller_3_test.rb:1593]: I don't see the indentation issue you're talking about?

File: app/models/bot/alegre.rb (hunk 15 above)
Comment (DGaffney) [reply app/models/bot/alegre.rb:156]: ok!

File: app/models/bot/alegre.rb (hunk 17 above)
Comment (DGaffney) [reply app/models/bot/alegre.rb:20]: lol fair point

File: app/models/project_media.rb (hunk 18 above)
Comment (DGaffney) [reply app/models/project_media.rb:53]: Ah -good catch - thank you!

**Comment by DGaffney** [comment]
Running a test on Travis now...

File: app/models/bot/alegre.rb [hunk 19]
```diff
@@ -115,63 +163,89 @@ def self.request_api(method, path, params = {})
     http.use_ssl = uri.scheme == 'https'
     begin
       response = http.request(request)
-      JSON.parse(response.body)
+      puts response.body
```
Comment (caiosba) [inline app/models/bot/alegre.rb:166]: Please remove this line :D

Comment (caiosba) [reply app/models/bot/alegre.rb:166]: Or replace by `Rails.logger.info`, with some prefix if it's import to log this output
